### PR TITLE
Run E2E tests inside a kind cluster

### LINKED ISSUE #250: Use a DIND test jig for automation
As part of the upstream testing commons work we should eventually be able to pull in a integration test framework that spins a DIND cluster that allows us to exercise larger integration pieces that are manual today.  

This is a long term effort.

### PR DESCRIPTION
**Pull request by stevesloka**
**What this PR does / why we need it**:

This PR now builds a kind cluster inside travis, runs the E2E tests and checks for a valid result. We'll need to see if this should be a 2nd job or have a bot only run occasionally since it will take some time to run. 

**Which issue(s) this PR fixes**
- Fixes #250 

Signed-off-by: Steve Sloka <slokas@vmware.com>

**Release note**:
```
release-note
```

### CODE REVIEW

**Comment by timothysc** [comment]
This makes me so happy!  Thx @stevesloka!

File: travis-ci.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+  echo "results not ready..."
+  false
+}
+
+echo "|---- Creating new Sonobuoy run..."
+$SONOBUOY_CMD run
+waitForDeployment
+
+while [ !checkForResults ]
```
Comment (johnSchnake) [inline travis-ci.sh]: We'd want to be careful that this wouldn't end in an infinite loop if a deploy failed.

File: travis-ci.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+if [ -z "$TRAVIS" ]; then
+    echo "this script is intended to be run only on travis" >&2
+    exit 1
+fi
+
+KUBECTL_WAIT_CMD="kubectl wait --timeout=5m --for=condition=Ready pod/sonobuoy -n heptio-sonobuoy"
+SONOBUOY_CMD=./sonobuoy
+SONOBUOY_SUB_CMD=run -m Quick
+
+function waitForDeployment() {
```
Comment (johnSchnake) [inline travis-ci.sh]: I have a PR nearly ready (maybe putting it up tomorrow) which would introduce the `--wait` options to run and delete commands. This makes scripting this much easier.

File: travis-ci.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      $KUBECTL_WAIT_CMD
+    fi
+  done
+}
+
+function checkForResults() {
+  echo "|---- Looking for results..."
+
+  # Attempt to get status
+  $SONOBUOY_CMD retrieve
```
Comment (johnSchnake) [inline travis-ci.sh]: So the fear as of now is that you can get a `retrieve` that generates a .tar.gz but that is corrupted because it was grabbed before the writer was done.

That gets solved if https://github.com/heptio/sonobuoy/pull/584 merges.

File: travis-ci.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  $SONOBUOY_CMD status
+  sleep 30
+done
+
+echo "|------ Results found, validating output..."
+
+for i in *.gz; do
+  mkdir -p results
+  tar -xvf $i -C results
```
Comment (johnSchnake) [inline travis-ci.sh]: Does the environment on travis get cleaned up properly behind the scenes, or do we need to handle that too?

File: travis-ci.sh [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  grep -q "Test Suite Passed" results/plugins/e2e/results/e2e.log
+  
+  if [ $? -eq 0 ]
+  then
+    echo "Success!"
+    exit 0
+  else
+    echo "Failure: E2E tests failed" >&2
```
Comment (johnSchnake) [inline travis-ci.sh]: Not sure the file retention via jenkins here, but if we can't grab the tar then we need to be sure to tail/echo the end of the file where all the failed tests are listed otherwise we'll be perpetually frustrated by not knowing which tests failed.

**Comment by stevesloka** [comment]
I think this is good now @johnSchnake! We'll need to come up with a way to make the full suite customizable. Maybe for each merge to master?

File: kind-config.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# three node (two workers) cluster config
+kind: Config
+apiVersion: kind.sigs.k8s.io/v1alpha2
+nodes:
+- role: control-plane
+- role: worker
+  replicas: 3
```
Comment (johnSchnake) [inline kind-config.yaml]: Out of sync from the comment; do we want 3 workers or 2?

File: travis-ci.sh [hunk 7]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+# Don't fail silently when a step doesn't succeed
+set -e
+
+# if [ -z "$TRAVIS" ]; then
```
Comment (johnSchnake) [inline travis-ci.sh]: Do we want this uncommented or removed?

File: travis-ci.sh [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+# fi
+
+make container deploy_kind
+
+echo "|---- Creating new Sonobuoy run/waiting for results..."
+./sonobuoy run --image-pull-policy=IfNotPresent -m Quick --wait & PID=$!
+
+# While process is running...
+while kill -0 $PID 2> /dev/null; do 
+    printf  "."
```
Comment (johnSchnake) [inline travis-ci.sh]: You started off the conversation about this in slack and the idea of just outputting a stream of `.` seems the most simple answer. I opened #607 regarding this.

File: travis-ci.sh [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+done
+
+echo "|---- Retrieving results..."
+./sonobuoy retrieve
+
+echo "|---- Results found, validating output..."
+for i in *.gz; do
+mkdir -p results
+tar -xvf $i -C results
+grep -q "Test Suite Passed" results/plugins/e2e/results/e2e.log
```
Comment (johnSchnake) [inline travis-ci.sh]: This section for getting and validating output can be reduced to something like:
```
outFile=$(./sonobuoy retrieve)
results=$(./sonobuoy e2e $outFile)
echo $results
exitCode=$($results | grep 0)
exit $exitCode
```

This is of course based on the idea that https://github.com/heptio/sonobuoy/pull/598 merges (any minute after tests go green).

That has the benefit that we also are 
1) dogfooding our own command
2) listing the failed tests, not just reporting failure or not (I know right now this is quick mode so its just one test, but you get the point)
3) less coupled to the text output by ginkgo for passed/failed
4) avoid some of the file creation/manipulation that can sometimes (in my experience) have odd edge cases

File: kind-config.yaml (hunk 6 above)
Comment (stevesloka) [reply kind-config.yaml]: 3 is a typical cluster that gives us enough workers I think for a test.

File: travis-ci.sh (hunk 7 above)
Comment (stevesloka) [reply travis-ci.sh]: I'll remove, I find it handy to kick it all off locally.

File: travis-ci.sh (hunk 8 above)
Comment (stevesloka) [reply travis-ci.sh]: I didn't want to wait for that so this is the easy fix. Showing output I think in the end game is best done with showing the total number of tests and which one is currently getting executed.

File: travis-ci.sh (hunk 9 above)
Comment (stevesloka) [reply travis-ci.sh]: We need to look at `sonobuoy e2e outfile`, it's not super helpful which is why I avoided (and your PR was still in flight).

File: travis-ci.sh (hunk 8 above)
Comment (johnSchnake) [reply travis-ci.sh]: Yea absolutely; I should have specified I didnt want that to be a blocker but just wanted to start the ticket since we had talked about it.

File: kind-config.yaml (hunk 6 above)
Comment (johnSchnake) [reply kind-config.yaml]: 👍 thats fine; I just didnt want the comment there saying 2 if we were setting 3.

**Comment by stevesloka** [comment]
@johnSchnake all cleaned up with comments. Please take a look, thanks! =)

File: travis-ci.sh [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+echo "|---- Creating new Sonobuoy run/waiting for results..."
+./sonobuoy run --image-pull-policy=IfNotPresent -m Quick --wait
+
+outFile=$(./sonobuoy retrieve)
+results=$(./sonobuoy e2e $outFile)
+
+echo $results
+
+if echo $results | grep --quiet "failed tests: 0"; then
```
Comment (aojea) [inline travis-ci.sh]: This is not enough condition to declare the test OK.
I have the same result in a failed deployment 
````
{"level":"info","msg":"Collecting Pod Logs...","time":"2019-03-05T12:35:40Z"}
{"level":"error","msg":"the server rejected our request for an unknown reason (get pods sonobuoy-e2e-job-27d5ca4105034c28)","
time":"2019-03-05T12:35:40Z"}
```

and failed tests :0 but the problem that it didn't run any test
```
sonobuoy e2e 201903051232_sonobuoy_7cd48d20-e974-42dd-8a4b-1dbcf0ae3bd2.tar.gz
failed tests: 0
```
